projects/extension/src/background/index.ts

```diff
@@ -191,6 +193,55 @@ window.uiInterface = {
       }
     }
   },
+  // Based on the chain provided, it returns the default bootnodes of given chain
+  // bootnodes are retrieved from the chainspecs that exist in the extension.
+  getDefaultBootnodes: (chain: string): string[] => {
+    if (chain === "polkadot") return polkadot.bootNodes
+    if (chain === "ksmcc3") return ksmcc3.bootNodes
+    if (chain === "westend2") return westend2.bootNodes
+    if (chain === "rococo_v2_2") return rococo_v2_2.bootNodes
+    return []
+  },
+  // Sends an message through rpc to smoldot in order to validate the given bootnode.
+  // This will return to the caller the response once received from rpc.
+  // Message's id is prefixed with "extension:" in order to be differentiated from the rest
+  // and be able to reply specific message at nextSandboxMessage to the caller.
+  updateBootnode: (chain: string, bootnode: string, add: boolean): void => {
+    if (!add) {
+      // remove bootnode from localstorage if it already exists
+      chrome.storage.local.get(["bootNodes_".concat(chain)], (result) => {
+        const res = [...result["bootNodes_".concat(chain)]]
+        const idx = res.findIndex((a) => a === bootnode)
+        if (idx > -1) {
+          res.splice(idx, 1)
+          chrome.storage.local.set({ ["bootNodes_".concat(chain)]: [...res] })
+        }
+      })
+    } else {
+      // if it does not exist, then send for validation
+      if (manager.state !== "ready") return
+      const nextId = manager.manager.nextRpcRqId()
+      chrome.storage.local.set({
+        ["customBootnode"]: {
+          id: nextId,
+          chain,
+          bootnode,
+          result: -1,
+        },
+      })
+      manager.manager.sandboxMessage(null, {
+        chainId: chain,
+        type: "rpc",
+        origin: "substrate-connect-client",
+        jsonRpcMessage: JSON.stringify({
+          id: "extension:" + nextId,
+          jsonrpc: "2.0",
+          method: "sudo_unstable_p2pDiscover",
+          params: [bootnode],
+        }),
+      })
```

Comment:
So you're using the local storage as if it was a variable, basically.
What if the extension shuts down while the request is in progress. The next time it loads you'll still have the item in there, except that now it doesn't correspond to any request.

Please instead just add an asynchronous method to the `ConnectionManagerWithHealth` that sends a request, and returns the response to the request.